Working with dev mode in the Eclipse IDE (#2102)
OpenLiberty/blogs: posts/2022-03-11-dev-mode-with-eclipse.adoc

link:{url-prefix}/docs/latest/development-mode.html[Liberty dev mode] enables you to quickly and iteratively develop cloud-native Java applications with the latest Jakarta EE and MicroProfile technologies. Dev mode watches your project for file changes and provides hot reload and deployment, on demand testing, and debugger support. You can work with dev mode from anywhere, whether it's from a terminal with a text editor or your favorite IDE. https://marketplace.eclipse.org/content/ibm-liberty-developer-tools[Liberty Developer Tools] is an Eclipse plugin that provides helpful server configuration editor views. In the following sections, we detail the current best practices for using dev mode, and optionally Liberty Developer Tools, to develop your application within the Eclipse IDE.

Check out a demo of working with Liberty dev mode in the Eclipse IDE:

++++
<iframe width="560" height="315" src="https://www.youtube.com/embed/613VBYdk6f8" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
++++

== Working with dev mode in the Eclipse IDE with Liberty Developer Tools

> If you have Liberty Developer Tools installed in your Eclipse workspace, the server configuration editor views are helpful as you modify your `server.xml` and other configuration files. You can easily add new server configuration elements and features to your `server.xml` file.

Reviewer: ```suggestion
If you have Liberty Developer Tools installed in your Eclipse workspace, the **Server Configuration** editor is  helpful as you modify your `server.xml` and other configuration files. You can easily add new server configuration elements and features to your `server.xml` file.
```

image::/img/blog/liberty-dev-tools-server-config-view.png[Liberty Developer Tools server configuration view for the Eclipse IDE,width=70%,align="center"]

> Dev mode handles a lot of the heavy lifting for developers who use the Liberty runtime. With dev mode, you no longer need to manage your Liberty runtime instances through the **Servers** view of Liberty Developer Tools.

Reviewer: ```suggestion
However, dev mode handles a lot of the heavy lifting for developers who use the Liberty runtime. With dev mode, you no longer need to manage your Liberty runtime instances through the **Servers** view of Liberty Developer Tools. You can disable the **Servers** view from the Liberty Developer Tools **Preferences** menu.
```
```suggestion
Dev mode handles a lot of the heavy lifting for developers who use the Liberty runtime. With dev mode, you no longer need to manage your Liberty runtime instances through the **Servers** view of Liberty Developer Tools.
```

> Disable Liberty Developer Tools' integration for Maven and Gradle projects:

Reviewer: when you disable the integration, do you also lose access to the Server Configuration view? or can you continue to use that to make changes to the server.xml that dev mode will automatically detect?

Author: You can continue to use the Server configuration view and dev mode will pick up the changes to the server.xml file

---
layout: post
title: "Rapid cloud-native development in the Eclipse IDE with Liberty dev mode"
# Do NOT change the categories section
categories: blog
author_picture: https://avatars3.githubusercontent.com/kathrynkodama
author_github: https://github.com/kathrynkodama
seo-title: Rapid cloud-native development in the Eclipse IDE with Liberty dev mode - OpenLiberty.io
seo-description: Liberty dev mode enables developers to rapidly develop cloud-native Java applications using MicroProfile and Jakarta EE in the Eclipse IDE.
blog_description: "Liberty dev mode enables developers to rapidly develop cloud-native Java applications using MicroProfile and Jakarta EE in the Eclipse IDE."

> open-graph-image: https://openliberty.io/img/twitter_card.jpg

Reviewer: ```suggestion
open-graph-image: https://openliberty.io/img/blog/liberty-dev-tools-server-config-view.png
```

---
= Rapid cloud-native development in the Eclipse IDE with Liberty dev mode
Kathryn Kodama <https://github.com/kathrynkodama>
:imagesdir: /
:url-prefix:
:url-about: /

> The Eclipse IDE (integrated development environment) continues to be one of the most popular IDEs among Java developers. In this post we showcase how Liberty dev mode and the Liberty Developer Tools Eclipse plugin ensure a seamless development experience for cloud-native Java developers.

Reviewer: ```suggestion
The Eclipse integrated development environment (IDE) is one of the most popular IDEs among Java developers. In this post, we showcase how Liberty dev mode and the Liberty Developer Tools Eclipse plugin ensure a seamless development experience for cloud-native Java developers.
```

> https://openliberty.io/docs/latest/development-mode.html[Liberty dev mode] enables developers to quickly and iteratively develop cloud-native Java applications with the latest Jakarta EE and MicroProfile technologies. Dev mode will watch your project for file changes and provide hot reload and deployment, on demand testing, and debugger support. You can work with dev mode from anywhere, whether it's from a terminal with a text editor or your favorite IDE. https://marketplace.eclipse.org/content/ibm-liberty-developer-tools[Liberty Developer Tools] is an Eclipse plugin that provides helpful server configuration editor views. In the following sections we detail the current best practice for using dev mode, and optionally Liberty Developer Tools, to develop your application within the Eclipse IDE.

Reviewer: ```suggestion
link:{url-prefix}/docs/latest/development-mode.html[Liberty dev mode] enables you to quickly and iteratively develop cloud-native Java applications with the latest Jakarta EE and MicroProfile technologies. Dev mode watches your project for file changes and provides hot reload and deployment, on demand testing, and debugger support. You can work with dev mode from anywhere, whether it's from a terminal with a text editor or your favorite IDE. https://marketplace.eclipse.org/content/ibm-liberty-developer-tools[Liberty Developer Tools] is an Eclipse plugin that provides helpful server configuration editor views. In the following sections, we detail the current best practices for using dev mode, and optionally Liberty Developer Tools, to develop your application within the Eclipse IDE.
```

Check out a demo of working with Liberty dev mode in the Eclipse IDE:

++++
<iframe width="560" height="315" src="https://www.youtube.com/embed/613VBYdk6f8" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
++++

> == Working with dev mode in the Eclipse IDE with Liberty Developer Tools installed

Reviewer: ```suggestion
== Working with dev mode in the Eclipse IDE with Liberty Developer Tools
```

If you have Liberty Developer Tools installed in your Eclipse workspace, the server configuration editor views can be helpful as you modify your `server.xml` and other configuration files.

image::/img/blog/liberty-dev-tools-server-config-view.png[Liberty Developer Tools server configuration view for the Eclipse IDE,width=70%,align="center"]

> Dev mode handles a lot of the heavy lifting for developers when using the Liberty runtime. With dev mode, you no longer need to manage your Liberty instances through the "Servers" view of Liberty Developer Tools.

Reviewer: ```suggestion
Dev mode handles a lot of the heavy lifting for developers who use the Liberty runtime. With dev mode, you no longer need to manage your Liberty instances through the **Servers** view of Liberty Developer Tools.
```

> Disable Liberty Developer Tools' integration for Maven and Gradle projects:

Reviewer: this is a little confusing- the section is about working with LDT in eclipse and says the Servers view is helpful, but the main thing it tells you is how to disable the integration. Are you disabling all of LDT or just the servers view? Its not clear

Reviewer: Im confused about how LDT fit into this post- there doesn't seem to be anything specific to LDT except the servers view and its not clear how this relates to dev mode or Eclipse

Author: I am trying to convey that LDT has helpful server configuration editor views that appear when editing server.xml files. However, with dev mode it is recommended that the "Servers" view contributed by LDT is disabled. I can see it is confusing between the two: server configuration editor views vs "Servers" view

Author: I've updated the section slightly in an attempt to clarify between the two, please let me know what you think.

* For Maven projects: **Window -> Preferences -> Maven -> Liberty Integration** drop-down to "No".

* For Gradle projects: **Window -> Preferences -> Gradle -> Liberty Integration** drop-down to "No".

image::/img/blog/liberty-dev-tools-disable-servers-view.png[Liberty Developer Tools disabling the Servers view for the Eclipse IDE,width=70%,align="center"]

== Setting up the project

> 

Reviewer: does this part have anything to do with Eclipse, dev mode, or LDT? Or is it just generally what tyou have to do for any Liberty Maven/Gradle projects?

Author: Generally what you have to do for Maven/Gradle projects but also a critical step in order to use dev mode. I've added a sentence at the beginning of the section to clarify.

For Maven projects, specify the Liberty Maven Plugin in your project's `pom.xml` file:
[source,xml]
----
    <plugin>
        <groupId>io.openliberty.tools</groupId>
        <artifactId>liberty-maven-plugin</artifactId>
        <version>3.5.1</version>
    </plugin>
----

For Gradle projects, specify the Liberty Gradle Plugin in your project's `build.gradle` file:
[source,groovy]
----
buildscript {
    repositories {
        mavenCentral()
    }
    dependencies {
        classpath 'io.openliberty.tools:liberty-gradle-plugin:3.3'
    }
}
----

> == Interacting with dev mode

Reviewer: shouldn't have two headings back to back- I thibnk you could remove the "Interacting..." heading since all the content is under subheadings and just add "with Dev Mode" to the subheadings

=== Starting dev mode

> Start dev mode through the Eclipse IDE terminal (Navigate to your project in the package explorer -> Right-click -> "Show in Terminal") or through your operating system's terminal with the following commands:

Reviewer: ```suggestion
You can start dev mode through the Eclipse IDE terminal by finding your project in the package explorer  and right-clicking **Show in Terminal**. Alternatively, you can start it from your operating system terminal with one of the following commands:
```

* Maven: `mvn liberty:dev`
* Gradle: `gradle libertyDev`

image::/img/blog/dev-mode-eclipse-ide-terminal.png[Starting Liberty dev mode through the Eclipse IDE terminal,width=70%,align="center"]

> === Running tests

Reviewer: ```suggestion
=== Running tests in dev mode
```

You can run tests on demand by pressing Enter in the terminal running dev mode.

> 

Reviewer: can you run tests from Eclipse?

Author: This may be the Eclipse IDE terminal or an operating system terminal referenced in the "Starting dev mode" section above.

=== Attaching a debugger

> You can attach a debugger to the running server to step through your code at any time. In Eclipse you can create a debug configuration by selecting "Run" -> "Debug Configurations..." -> "Remote Java application". The default port for debugging is `7777`.

Reviewer: ```suggestion
You can attach a debugger to the running server to step through your code at any time. In Eclipse, you can create a debug configuration by selecting **Run** -> **Debug Configurations...** -> **Remote Java application**. The default port for debugging is `7777`.
```

image::/img/blog/dev-mode-eclipse-ide-debug-config.png[Liberty debug configuration for the Eclipse IDE,width=70%,align="center"]

=== Stopping dev mode

You can quit dev mode at any time by pressing CTRL+C in the terminal, or type 'q' and press Enter.

> 

Reviewer: can you do this from Eclipse?

Author: This may be the Eclipse IDE terminal or an operating system terminal referenced in the "Starting dev mode" section above.
